pkg/controllers/node/suite_test.go

```diff
@@ -285,6 +286,27 @@ var _ = Describe("Controller", func() {
 			node = ExpectNodeExists(ctx, env.Client, node.Name)
 			Expect(node.DeletionTimestamp.IsZero()).To(BeFalse())
 		})
+		It("should requeue reconcile if node is empty, but not past emptiness TTL", func() {
+			provisioner.Spec.TTLSecondsAfterEmpty = ptr.Int64(30)
+			currentTime := time.Now()
+			injectabletime.Now = func() time.Time { return currentTime } // injectabletime.Now() is called multiple times in function being tested.
+			emptinessTime := injectabletime.Now().Add(-10 * time.Second)
+			// Emptiness timestamps are first formatted to a string friendly (time.RFC3339) (to put it in the node object)
+			// and then eventually parsed back into time.Time when comparing ttls. Repeating that logic in the test.
+			emptinessTimestamp, _ := time.Parse(time.RFC3339, emptinessTime.Format(time.RFC3339))
+			expectedrequeueTime := emptinessTimestamp.Add(time.Duration(30 * time.Second)).Sub(injectabletime.Now()) // we should requeue in ~20 seconds.
+			node := test.Node(test.NodeOptions{ObjectMeta: metav1.ObjectMeta{
+				Finalizers: []string{v1alpha5.TerminationFinalizer},
+				Labels:     map[string]string{v1alpha5.ProvisionerNameLabelKey: provisioner.Name},
+				Annotations: map[string]string{
+					v1alpha5.EmptinessTimestampAnnotationKey: emptinessTime.Format(time.RFC3339),
+				}},
+			})
+			ExpectCreated(ctx, env.Client, provisioner, node)
+			ExpectReconcileSucceededWithResult(ctx, controller, client.ObjectKeyFromObject(node), reconcile.Result{Requeue: true, RequeueAfter: expectedrequeueTime})
+			node = ExpectNodeExists(ctx, env.Client, node.Name)
+			Expect(node.DeletionTimestamp.IsZero()).To(BeTrue())
+		})
```

Comment:
Would you be able to add another unit test that replicates and targets the issue this was opened for?

The steps would be the following:
1) Create node with pod on it
2) Set Provisioner.TTLAfterEmpty to 30
3) Delete pod from node
4) Check Node.TTLAfterEmpty is set
5) Add another pod
6) Check Node.TTLAfterEmpty unset
7) remove second pod
8) Check Node.TTLAfterEmpty is set
9) Simulate 30 seconds of time passing
10) check that node deletion timestamp is nonZero

Reply:
I don't think that specific scenario is suitable for a unit test IMO. I think we should be testing what our code expects to do (i.e set a RequeueAfter with value foo) rather than test every possible permutation.